# Home Credit — широкий набор агрегатов и ограниченный поиск

[Home Credit Default Risk](https://www.kaggle.com/competitions/home-credit-default-risk): бинарная классификация с `roc_auc`. Код агрегирует вспомогательные таблицы до одной строки на заявку и использует подвыборку `APP_SAMPLE`. Библиотека получает подготовленную широкую матрицу; агрегация выполняется в ноутбуке.

Эксперимент сравнивает исходные модели и расширенный поиск с FS/HPO. Плотность пропусков, избыточность и доступность агрегатов на момент заявки требуют анализа; само наличие большого числа признаков не гарантирует выигрыш от FS.

**Срез результатов: 9 сентября 2026.** Итоговые таблицы используют связанные отчёты контрольных прогонов: seed 42, четыре native/BLAS-потока, кеш выключен, HPO до пяти trials на семейство при его включении, отправки Kaggle выключены. Встроенные outputs относятся к собственному запуску ячеек; контрольные режимы baseline/fast запускались отдельным runner. Код ячеек сохраняет исходные бюджеты и не включает `SearchConfig`. Просмотренный holdout служит диагностикой. [Границы сравнения и независимой приёмки](README.md#контрольные-прогоны-и-интерпретация).

In [1]:
import json
import logging
import os
import shutil
import subprocess
import time
from pathlib import Path

import numpy as np
import pandas as pd
from IPython.display import Markdown

from honestml import (
    AutoML,
    CVConfig,
    FeatureSelectionConfig,
    HPOConfig,
    render_report,
    save_run_report,
)

SEED = 42
APP_SAMPLE = (
    150_000  # model on a stratified-by-chance subset to keep the run light; None = all 307k
)
DATA = Path("data/home-credit")
APP = DATA / "home-credit-default-risk"
RESULTS = Path("results/home-credit")
RESULTS.mkdir(parents=True, exist_ok=True)
KAGGLE = (
    os.environ.get("KAGGLE_BIN")
    or shutil.which("kaggle")
    or str(Path.home() / ".local" / "bin" / "kaggle")
)

logging.basicConfig(format="%(levelname)s %(name)s: %(message)s")
logging.getLogger("honestml").setLevel(logging.INFO)

In [2]:
if not (APP / "application_train.csv").exists():
    DATA.mkdir(parents=True, exist_ok=True)
    subprocess.run(
        [
            KAGGLE,
            "datasets",
            "download",
            "-d",
            "megancrenshaw/home-credit-default-risk",
            "-p",
            str(DATA),
            "--unzip",
        ],
        check=True,
    )

In [3]:
def aggregate(filename, prefix):
    """Collapse one auxiliary table to one row per SK_ID_CURR (numeric mean/max/min/sum + count)."""
    t = pd.read_csv(DATA / filename)
    drop = [c for c in ("SK_ID_PREV", "SK_ID_BUREAU") if c in t.columns]
    num = t.drop(columns=drop).select_dtypes("number")
    a = num.groupby("SK_ID_CURR").agg(["mean", "max", "min", "sum"])
    a.columns = [f"{prefix}_{col}_{stat}" for col, stat in a.columns]
    a[f"{prefix}_count"] = t.groupby("SK_ID_CURR").size()
    return a


app = pd.read_csv(APP / "application_train.csv")
app["DAYS_EMPLOYED"] = app["DAYS_EMPLOYED"].replace(365243, np.nan)  # documented sentinel
app["DAYS_EMPLOYED_PERC"] = app["DAYS_EMPLOYED"] / app["DAYS_BIRTH"]
app["INCOME_CREDIT_PERC"] = app["AMT_INCOME_TOTAL"] / app["AMT_CREDIT"]
app["ANNUITY_INCOME_PERC"] = app["AMT_ANNUITY"] / app["AMT_INCOME_TOTAL"]
app["EXT_SOURCES_MEAN"] = app[["EXT_SOURCE_1", "EXT_SOURCE_2", "EXT_SOURCE_3"]].mean(axis=1)

# aggregate the auxiliary tables to the applicant grain (each full table is read once, then freed)
for filename, prefix in [
    ("bureau.csv", "BURO"),
    ("previous_application.csv", "PREV"),
    ("POS_CASH_balance.csv", "POS"),
    ("installments_payments.csv", "INS"),
    ("credit_card_balance.csv", "CC"),
]:
    app = app.join(aggregate(filename, prefix), on="SK_ID_CURR")

if APP_SAMPLE is not None and len(app) > APP_SAMPLE:
    app = app.sample(APP_SAMPLE, random_state=SEED).reset_index(drop=True)
y = app.pop("TARGET")
X = app.drop(columns=["SK_ID_CURR"])
print(f"rows: {len(X)}, features: {X.shape[1]}, default rate: {y.mean():.3%}")
print(f"missing cells: {X.isna().mean().mean():.1%} of all values")
X.head()

rows: 150000, features: 377, default rate: 8.076%
missing cells: 29.9% of all values


,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,NAME_TYPE_SUITE,...,CC_CNT_INSTALMENT_MATURE_CUM_sum,CC_SK_DPD_mean,CC_SK_DPD_max,CC_SK_DPD_min,CC_SK_DPD_sum,CC_SK_DPD_DEF_mean,CC_SK_DPD_DEF_max,CC_SK_DPD_DEF_min,CC_SK_DPD_DEF_sum,CC_count
0,Cash loans,M,Y,N,2,207000.0,465457.5,52641.0,418500.0,Unaccompanied,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Cash loans,F,Y,Y,0,247500.0,1281712.5,48946.5,1179000.0,Unaccompanied,...,39.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,24.0
2,Cash loans,F,Y,N,0,202500.0,495000.0,39109.5,495000.0,Unaccompanied,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,Cash loans,F,N,Y,0,247500.0,254700.0,24939.0,225000.0,Unaccompanied,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,Cash loans,M,N,Y,0,112500.0,308133.0,15862.5,234000.0,Unaccompanied,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,25.0


## Level 1 — модели на полном наборе агрегатов

Конфигурация использует стратифицированный CV, outer holdout 20% и `calibrate="auto"`. Результат калибровки читается по отдельным метрикам; высокая ROC-AUC сама по себе не подтверждает точность вероятностей или пригодность кредитного решения.

In [4]:
model = AutoML(
    task="binary",
    metric="roc_auc",
    cv=CVConfig(outer_holdout=0.2, calibrate="auto"),
    random_state=SEED,
)
t0 = time.perf_counter()
model.fit(X, y)
print(f"level-1 fit took {(time.perf_counter() - t0) / 60:.1f} min")
report = model.run_report_
pd.DataFrame(report["leaderboard"])

INFO honestml.adapters.reader: auto-typing column=CNT_CHILDREN dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_MOBIL dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_EMP_PHONE dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_WORK_PHONE dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_CONT_MOBILE dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_PHONE dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_EMAIL dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REGION_RATING_CLIENT dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REGION_RATING_CLIENT_W_CITY dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REG_REGION_NOT_LIVE_REGION dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REG_REGION_NOT_WORK_REGION dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=LIVE_REGION_NOT_WORK_REGION dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REG_CITY_NOT_LIVE_CITY dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REG_CITY_NOT_WORK_CITY dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=LIVE_CITY_NOT_WORK_CITY dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_2 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_3 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_4 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_5 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_6 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_7 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_8 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_9 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_10 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_11 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_12 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_13 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_14 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_15 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_16 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_17 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_18 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_19 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_20 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_21 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml: stage key=run stage=selection elapsed=516.9s


WARNING honestml.adapters.boosting: boosting 'catboost' trained without early stopping; leaderboard comparison may favor overfit settings


INFO honestml: stage key=run stage=refit elapsed=51.6s


INFO honestml: stage key=run stage=refit elapsed=60.6s


INFO honestml: stage key=run stage=finalize elapsed=60.6s


level-1 fit took 10.5 min


,model_id,score,rank
0,catboost,0.771815,1
1,linear,0.764319,2
2,lightgbm,0.764153,3
3,baseline,0.499955,4


In [5]:
selection = next(e["score"] for e in report["leaderboard"] if e["model_id"] == report["winner"])
print(f"winner          : {report['winner']}")
print(f"equivalence band: {report['band']['member_ids']}")
print(f"selection score : {selection:.4f}   (out-of-fold roc_auc)")
print(f"holdout score   : {report['holdout_score']:.4f}   (untouched 20%, scored once)")
print(f"optimism        : {selection - report['holdout_score']:+.4f}")
cal = model.calibration_
if cal:
    print(
        f"calibration     : Brier {cal.get('brier_raw'):.4f} -> {cal.get('brier_calibrated'):.4f}, "
        f"ECE {cal.get('ece_raw'):.4f} -> {cal.get('ece_calibrated'):.4f} ({cal.get('method')})"
    )
save_run_report(report, RESULTS)
md_path = render_report(report, RESULTS, fmt="md")
Markdown(md_path.read_text(encoding="utf-8"))

winner          : catboost
equivalence band: ['catboost']
selection score : 0.7718   (out-of-fold roc_auc)
holdout score   : 0.7695   (untouched 20%, scored once)
optimism        : +0.0023
calibration     : Brier 0.0669 -> 0.0669, ECE 0.0020 -> 0.0008 (auto)


# AutoML run report

## Run

| | |
|---|---|
| task | binary |
| metric | roc_auc |
| winner | catboost |
| holdout_score | 0.769505 |
| honestml_version | 1.1.0 |
| run_fingerprint | bd67bd2e95c09dfd49ac0de37976f076fef5dcca01c2a9ab2a5b13f5e7940fbf |
| significance | bootstrap |
| preset | n/a |

## Equivalence band

| | |
|---|---|
| members | catboost |
| width | 1 |
| unstable | False |
| winner_by_tiebreak | False |

## Budget

| | |
|---|---|
| mode | none |
| exhausted | False |
| exhausted_by | n/a |
| skipped | n/a |

## Serving

| | |
|---|---|
| finalize | True |
| shipped_on | all |
| outer_holdout | 0.2 |

## Leaderboard

| rank | model | score |
|---|---|---|
| 1 | catboost (winner) | 0.771815 |
| 2 | linear | 0.764319 |
| 3 | lightgbm | 0.764153 |
| 4 | baseline | 0.499955 |

## Timings (s)

| stage | elapsed |
|---|---|
| run.selection | 516.9 |
| run.refit | 60.6 |
| run.finalize | 60.6 |

## Resolved config

```json
{
  "seed": 42,
  "cv": {
    "scheme": "stratified",
    "n_splits": 5,
    "n_test": 1,
    "n_es": 1,
    "purge": 0,
    "embargo": 0,
    "period": null,
    "period_size": null,
    "step_periods": null,
    "purge_delta": null,
    "embargo_delta": null,
    "max_train_periods": null,
    "max_train_size": null,
    "weighting": "pooled",
    "calibrate": "auto",
    "selection": "raw",
    "refinement_min_oof": 2000,
    "outer_holdout": 0.2
  },
  "budget": {
    "mode": "none",
    "time_budget_s": null,
    "n_trials": null,
    "memory_limit_mb": null
  },
  "fe": {
    "target_encoding": false,
    "te_smoothing": 10.0,
    "frequency_encoding": false,
    "intersections": false,
    "max_pairs": 50
  },
  "fs": null,
  "hpo": null,
  "ensemble": null,
  "significance": "bootstrap",
  "run_mode": "full",
  "model_types": [
    "catboost",
    "lightgbm"
  ]
}
```


## Level 2 — FS и HPO

Код включает `importance` FS, `preset="best"` и `HPOConfig(n_trials=15, timeout_s=600)`. Контрольные прогоны ограничивали HPO пятью trials на семейство. Итогом после выполнения FS/HPO может стать широкий контроль с исходной фабрикой.

`feature_selection=None` в итоговом отчёте не означает нулевую стоимость FS: выполненные fit и стадии проверяются по журналу работы.

In [6]:
model_full = AutoML(
    task="binary",
    metric="roc_auc",
    preset="best",
    hpo=HPOConfig(n_trials=15, timeout_s=600),
    feature_selection=FeatureSelectionConfig(strategy="importance", cutoff="auto"),
    cv=CVConfig(outer_holdout=0.2, calibrate="auto"),
    random_state=SEED,
)
t0 = time.perf_counter()
model_full.fit(X, y)
print(f"level-2 fit took {(time.perf_counter() - t0) / 60:.1f} min")

INFO honestml.composition.presets: preset best applied: ['ensemble']


INFO honestml.adapters.reader: auto-typing column=CNT_CHILDREN dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_MOBIL dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_EMP_PHONE dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_WORK_PHONE dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_CONT_MOBILE dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_PHONE dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_EMAIL dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REGION_RATING_CLIENT dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REGION_RATING_CLIENT_W_CITY dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REG_REGION_NOT_LIVE_REGION dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REG_REGION_NOT_WORK_REGION dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=LIVE_REGION_NOT_WORK_REGION dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REG_CITY_NOT_LIVE_CITY dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=REG_CITY_NOT_WORK_CITY dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=LIVE_CITY_NOT_WORK_CITY dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_2 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_3 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_4 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_5 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_6 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_7 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_8 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_9 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_10 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_11 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_12 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_13 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_14 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_15 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_16 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_17 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_18 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_19 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_20 dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=FLAG_DOCUMENT_21 dtype=Int64 role=categorical reason=low_cardinality_int


WARNING honestml.application.feature_selection: feature selection kept 187 of 377 features


C:\Users\Admin\Documents\AutoML\.venv\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


WARNING honestml.adapters.boosting: boosting 'lightgbm' trained without early stopping; leaderboard comparison may favor overfit settings


INFO honestml: stage key=run stage=hpo elapsed=735.4s


INFO honestml: stage key=run stage=selection elapsed=1108.9s


INFO honestml: stage key=run stage=ensemble elapsed=95.7s


INFO honestml: stage key=run stage=refit elapsed=19.0s


INFO honestml: stage key=run stage=refit elapsed=24.2s


INFO honestml: stage key=run stage=finalize elapsed=24.2s


level-2 fit took 20.9 min


In [7]:
report_full = model_full.run_report_
sel_full = next(
    e["score"] for e in report_full["leaderboard"] if e["model_id"] == report_full["winner"]
)
fs = report_full["feature_selection"]
print(f"winner          : {report_full['winner']}")
print(f"selection score : {sel_full:.4f}   (level 1: {selection:.4f})")
print(
    f"holdout score   : {report_full['holdout_score']:.4f}   (level 1: {report['holdout_score']:.4f})"
)
print(f"optimism        : {sel_full - report_full['holdout_score']:+.4f}")
if fs is None:
    search = report_full.get("search") or {}
    print(f"features kept   : full feature set  (gate: {search.get('final_control')})")
else:
    print(
        f"features kept   : {fs.get('n_selected')} of {X.shape[1]}  (gate: {fs.get('no_selection_gate')})"
    )
print()
print("hpo     :", json.dumps(report_full["hpo"], default=str)[:400])
print("ensemble:", json.dumps(report_full["ensemble"], default=str)[:300])
pd.DataFrame(report_full["leaderboard"])

winner          : lightgbm
selection score : 0.7725   (level 1: 0.7718)
holdout score   : 0.7790   (level 1: 0.7695)
optimism        : -0.0065
features kept   : 187 of 377  (gate: selection_kept)

hpo     : {"backend": "optuna", "inner_cv": 3, "deterministic": false, "selection_oof_is_post_tuning": true, "tuned_on_full_feature_space": false, "tuned_on": "fs_subset", "cost_estimate_fits": 90, "tuned": {"catboost": {"chosen_params": {"depth": 4, "learning_rate": 0.19030368381735815, "iterations": 350, "l2_leaf_reg": 5.105903209394756, "subsample": 0.608233797718321, "one_hot_max_size": 63}, "inner_best
ensemble: {"applied": true, "method": "caruana", "member_ids": ["baseline", "catboost", "lightgbm", "linear"], "weights": {"baseline": 0.4579560155239327, "catboost": 0.06985769728331177, "lightgbm": 0.34152652005174644, "linear": 0.13065976714100905}, "gate_reason": "significant_improvement", "oof_delta": 0.


,model_id,score,rank
0,lightgbm,0.772528,1
1,catboost,0.769214,2
2,linear,0.762966,3
3,baseline,0.499955,4


## Подтверждённые результаты Home Credit

| Сценарий | Режим | Итоговая модель | Fit, с | Просмотренный holdout |
|---|---|---|---:|---:|
| Level 1 | baseline | catboost | 757.18 | 0.775250 |
| Level 2 | baseline | ансамбль (опорная lightgbm) | 863.13 | 0.779242 |
| Level 1 | final fast | catboost | 623.82 | 0.775250 |
| Level 2 | final fast | catboost | 981.62 | 0.775250 |

Источник: [final Home analysis](../docs/training-results.md#final-runs). Baseline относится к сравнительной серии; `final fast` — отдельный завершённый запуск со своим снимком кода. Исходная frozen fast-попытка завершила оба fit, но упала в отчётной ячейке; она не входит в primary comparison. Её статус не подменяется успешным final fast.

Оба final fast fit сохранили одиночный CatBoost, 377 признаков и viewed holdout ROC-AUC 0,7752502351. Во втором fit выполнены FS на 191 признаке и пять HPO trials; затем выбран `wide_control` с исходной фабрикой. Получены 40 fit без ошибок, но итоговое качество совпало с level 1.

Baseline level 2 применил ансамбль на 187 признаках и получил 0,7792418126: на 0,0039915775 выше final fast. В этой паре различаются итоговые процедуры, а final fast также занимает больше времени. Универсальный выигрыш FS/HPO или ускорение относительно прода не подтверждены. Эти прогоны предшествуют добавлению условного прогноза FS/HPO; независимая приёмка качества остаётся открытой.